### ISSUE TITLE
Installing a different version of a package that is a dependency of a dependency causes TypeScript errors

### ISSUE THREAD

**Issue by divmgl** [oven-sh/bun#4642]
### What version of Bun is running?

1.0.0+822a00c4d508b54f650933a73ca5f4a3af9a7983

### What platform is your computer?

Darwin 22.5.0 arm64 arm

### What steps can reproduce the bug?

```json
// package.json
{
  "dependencies": {
    "bullmq": "3.6.2",
    "ioredis": "5.3.0"
  }
}
```
```ts
import { Redis } from "ioredis";
import { Queue } from "bullmq";

const redis = new Redis();
const bullmq = new Queue("my-random-queue", {
  connection: redis, // Typescript error occurs here
});
```
```
$ bun i
$ bun --watch run index.ts
```

### What is the expected behavior?

TypeScript doesn't complain.

### What do you see instead?

Instead, I see:
```
Type 'Redis' is not assignable to type 'ConnectionOptions | undefined'.
  Type 'Redis' is missing the following properties from type 'Redis': on, once ts(2322)
(property) QueueBaseOptions.connection?: ConnectionOptions | undefined
```
This is completely wrong, given that `ConnectionOptions` is:

```ts
import type * as IORedis from 'ioredis';
interface BaseOptions {
    skipVersionCheck?: boolean;
}
export type RedisOptions = IORedis.RedisOptions & BaseOptions;
export type ClusterOptions = IORedis.ClusterOptions & BaseOptions;
export type ConnectionOptions = RedisOptions | ClusterOptions | IORedis.Redis | IORedis.Cluster;
export {};
```

**Comment by smerchek** [comment]
I'm also experiencing this on our large repo (4210 installs across 3286 packages).

To add some other examples that may help:

```
error TS2345: Argument of type 'ZodObject<{ token: ZodString; }, "strip", ZodTypeAny, { token: string; }, { token: string; }>' is not assignable to parameter of type 'ZodType<any, any, any>'.
  Property 'readonly' is missing in type 'ZodObject<{ token: ZodString; }, "strip", ZodTypeAny, { token: string; }, { token: string; }>' but required in type 'ZodType<any, any, any>'.
  
error TS2345: Argument of type 'import("/Users/scott/code/udisc/web-copy/node_modules/@react-stately/calendar/dist/types").RangeCalendarState' is not assignable to parameter of type 'import("/Users/scott/code/udisc/web-copy/node_modules/react-aria/node_modules/@react-aria/calendar/node_modules/@react-stately/calendar/dist/types").RangeCalendarState'.
  Property 'isValueInvalid' is missing in type 'import("/Users/scott/code/udisc/web-copy/node_modules/@react-stately/calendar/dist/types").RangeCalendarState' but required in type 'import("/Users/scott/code/udisc/web-copy/node_modules/react-aria/node_modules/@react-aria/calendar/node_modules/@react-stately/calendar/dist/types").RangeCalendarState'.

error TS2321: Excessive stack depth comparing types 'Thing' and 'Thing'.

vitest.config.ts(9,22): error TS2322: Type 'Plugin_2' is not assignable to type 'PluginOption'.
  Type 'import("/Users/scott/code/udisc/web-copy/node_modules/vite-tsconfig-paths/node_modules/vite/dist/node/index").Plugin' is not assignable to type 'import("/Users/scott/code/udisc/web-copy/node_modules/vite/dist/node/index").Plugin'.
    Types of property 'apply' are incompatible.
```

At least in part, I was trying to decipher the produced `yarn.lock` produced by `bun install -y` and I know there are least a few differences in the installed modules moving from `pnpm` because of `^` package references (bun didn't use our pnpm-lock to get exact dependencies). 

I tried locking down our dependencies so they more closely matched our pnpm-lock set, but seems like this issue is broader than that.

From yarn.lock (via `bun install -y`):

```

zod@3.20.6:
  version "3.20.6"
  resolved "https://registry.npmjs.org/zod/-/zod-3.20.6.tgz"
  integrity sha512-oyu0m54SGCtzh6EClBVqDDlAYRz4jrVtKwQ7ZnsEmMI9HnzuZFj8QFwAY1M5uniIYACdGvv0PBWPF2kO0aNofA==

zod@^3.11.6, zod@^3.17.10, zod@^3.21.4:
  version "3.22.2"
  resolved "https://registry.npmjs.org/zod/-/zod-3.22.2.tgz"
  integrity sha512-wvWkphh5WQsJbVk1tbx1l1Ly4yg+XecD+Mq280uBGt9wa5BKSWf4Mhp6GmrkPixhMxmabYY7RbzlwVP32pbGCg==
  
"@react-aria/calendar@3.0.5":
  version "3.0.5"
  resolved "https://registry.npmjs.org/@react-aria/calendar/-/calendar-3.0.5.tgz"
  integrity sha512-RIOwGYIwMizN/MAF5RkTb2ic9OJ0rJyR2VqqgtV3c7ADHNejzyLYMQmaalEFDUHS+AbvaXM1LCXdFBhSB8nf5w==
  dependencies:
    "@internationalized/date" "^3.0.2"
    "@react-aria/i18n" "^3.6.3"
    "@react-aria/interactions" "^3.13.1"
    "@react-aria/live-announcer" "^3.1.2"
    "@react-aria/utils" "^3.14.2"
    "@react-stately/calendar" "^3.0.5"
    "@react-types/button" "^3.7.0"
    "@react-types/calendar" "^3.0.5"
    "@react-types/shared" "^3.16.0"
    "@swc/helpers" "^0.4.14"

"@react-aria/calendar@^3.2.0":
  version "3.5.0"
  resolved "https://registry.npmjs.org/@react-aria/calendar/-/calendar-3.5.0.tgz"
  integrity sha512-n/VrAJlKmsNrBKCPMI3tmCqpV38+Cihtinzp3yp8DeL44WH3IJij6aftkEcI7pIVNajY6vYLS9BbjscvIdg+fw==
  dependencies:
    "@internationalized/date" "^3.5.0"
    "@react-aria/i18n" "^3.8.2"
    "@react-aria/interactions" "^3.18.0"
    "@react-aria/live-announcer" "^3.3.1"
    "@react-aria/utils" "^3.20.0"
    "@react-stately/calendar" "^3.4.0"
    "@react-types/button" "^3.8.0"
    "@react-types/calendar" "^3.4.0"
    "@react-types/shared" "^3.20.0"
    "@swc/helpers" "^0.5.0"

```

**Comment by sirenkovladd** [comment]
This is because `bullmq` uses `"ioredis": "^5.3.0"` and you use `"ioredis": "5.3.0"`
And Bun installs for `bullmq` -> `ioredis@5.3.2` and for you `ioredis@5.3.0`

Of course, the easiest way is to set in package.json `"ioredis": "^5.3.0"`

**Comment by smerchek** [comment]
Since `ioredis@5.3.0` satisifies `"ioredis": "^5.3.0"` from `bullmq`, shouldn't it only install `ioredis@5.3.0` instead of both versions?

**Comment by divmgl** [comment]
> Of course, the easiest way is to set in package.json "ioredis": "^5.3.0"

Yep, I can unpin the dependency in my repo and that would likely solve the problem but the point is that we _want_ our dependencies to be pinned.

> Since ioredis@5.3.0 satisifies "ioredis": "^5.3.0" from bullmq, shouldn't it only install ioredis@5.3.0 instead of both versions?

I'm unsure what the behavior is for `pnpm` and the like, but this is what makes the most sense to me. 

I want to stress that we have not run into this issue with _all_ other package managers (`npm`, `yarn` and `pnpm`), so the expectation is that the Bun package manager also behaves in the same way.

**Comment by ymichael** [comment]
Hi! wanted to chime in and mention that we at Figma are also experiencing this same issue trying to switch to `bun` as our `npm` replacement. I believe that this issue is currently singlehandedly blocking us from adopting `bun`, with which we've observed a significant speed up in our development tooling.

As an aside, thank you so much for the work on this! we're are all really excited to try this out!

**Comment by dylan-conway** [comment]
this is fixed in bun v1.0.11 with #6814
